### PR TITLE
VAU and the Brainpool; plus some Kotlin - a more lightweight approach to ECIES

### PR DESCRIPTION
**Pull request by tobias-schwerdtfeger**

### CODE REVIEW

File: _posts/2022-10-XX-ecies-vau.markdown [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+If you look for a copy & paste solution head over to the [E-Rezept Android App VAU implementation](https://github.com/gematik/E-Rezept-App-Android/blob/master/common/src/commonMain/kotlin/de/gematik/ti/erp/app/vau/Crypto.kt).
+You can find some unit tests for this implementation [here](https://github.com/gematik/E-Rezept-App-Android/blob/master/common/src/commonTest/kotlin/de/gematik/ti/erp/app/vau/CryptoTest.kt).
+
+---
+
+Have some ❤️, question, feedback or feel like 🤔 Reach out to me on https://twitter.com/t0bsdt or https://github.com/tobias-schwerdtfeger
+
+# About the author
+
+Tobias Schwerdtfeger is a software developer for some years on Android with a special enthusiasm towards exploring and realising a broad selection of UI/UX concepts;
```
[the remark this answers is not in the document]
Comment (tobias-schwerdtfeger) [reply _posts/2022-10-XX-ecies-vau.markdown]: a particular; oh yes I love this word

File: _posts/2022-10-XX-ecies-vau.markdown [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+title:  "VAU and the Brainpool; plus some Kotlin - a more lightweight approach to ECIES"
+date:   2022-10-XX XX:XX:XX +0200
+author: Tobias Schwerdtfeger
+categories: tech
+tags: ECIES VAU TEE Kotlin Java Crypto BouncyCastle Brainpool NIST
+---
+
+Probably you're here for an answer on how to implement the Elliptic Curve Integrated Encryption Scheme (ECIES) in
+something like Kotlin or Java for communicating with the trusted execution environment
+(TEE; or in german we like to call this VAU) of our e-prescription server.
```
[the remark this answers is not in the document]
Comment (tobias-schwerdtfeger) [reply _posts/2022-10-XX-ecies-vau.markdown]: Well IBM uses TEE in their source code/ readme

File: _posts/2022-10-12-ecies-vau.markdown [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+```kotlin
+val cipher = Cipher.getInstance("ECIES", BCProvider)
+cipher.init(Cipher.ENCRYPT_MODE, publicKey)
+cipher.doFinal(content)
+```
+
+After your initial excitement, I've to tell you the truth: The code above is not sufficient for our needs 😔
+
+**BUT!** If you only want to use ECIES with some defaults (like NIST curves) using Android, I can recommend to use [Tink](https://github.com/google/tink/blob/master/docs/JAVA-HOWTO.md#hybrid-encryption).
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:23]: to use ->

File: _posts/2022-10-12-ecies-vau.markdown [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+_Disclaimer: I'm not a crypto expert who can answer the ins and outs of cryptographic schemes or elliptic curves.
+I just know how to implement this stuff. And it works! Well, now, not in the beginning...
+So if you only want to survive the loveliness of using rarely used elliptic curves and not so easy to implement cipher suits, you're welcome!_
+
+## How do we got here?
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:29]: got -> get

File: _posts/2022-10-12-ecies-vau.markdown [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+So let's take some steps back and actually explain the concept of the TEE.
+Just a couple of years ago, the e-prescription was in the making.
+At that time it was clear, that the patient needed some form of autonomy over his or her own health information.
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:33]: -> time,
-> clear [remove comma]
his or her -> their

File: _posts/2022-10-12-ecies-vau.markdown [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+So what does nearly everyone have nowadays? Yes, a smartphone! Therefore, a mobile app was the way to go.
+With the limitations even modern smartphones have, we needed some lightweight and safe encryption process, while following the guidelines of the [BSI](https://www.bsi.bund.de).
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:35]: -> process [remove the comma]

File: _posts/2022-10-12-ecies-vau.markdown [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+(Yes, that's the point where Brainpool curves come into play 🤫)
+
+The actual concept of the TEE represents some kind of environment, where no one (e.g. an administrator) is able from the outside to read, or modify data during the runtime of the (server-) application.
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:38]: -> environment [remove the comma]
-> e.g.,
-> read [remove the comma]

File: _posts/2022-10-12-ecies-vau.markdown [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+Another point was to ensure that the payload of the communication between the client (E-Rezept App, PVS, AVS) and the actual destination on the server stays always encrypted, until it enters the TEE.
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:39]: stay always encrypted, -> always stays encrypted

File: _posts/2022-10-12-ecies-vau.markdown [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+A normal HTTPS encryption would be terminated at the first endpoint, like the load balancer.
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:40]: normal -> standard

File: _posts/2022-10-12-ecies-vau.markdown [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+In this way, the e-prescription server, as a centralized architecture, can handle health data safely.
+
+**But what about ECIES now?** It's just an encryption scheme. Every implementation you'll find, will differ from each other on some (maybe tiny) details.
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:44]: -> find [remove the comma]

File: _posts/2022-10-12-ecies-vau.markdown [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+And you'll have to deal with random data! This will always lead to "in 1 of 1000 test runs it will magically fail".
```
Comment (noheadx) [inline _posts/2022-10-12-ecies-vau.markdown:45]: -> runs,

**Review by noheadx — CHANGES_REQUESTED** [review]

File: _posts/2022-10-12-ecies-vau.markdown (hunk 4 above)
Comment (tobias-schwerdtfeger) [reply _posts/2022-10-12-ecies-vau.markdown:29]: For some reason, for me it sounds always better with "got" haha

**Pull request merged**